### ISSUE TITLE
Bot is not working sometimes

### ISSUE THREAD

**Issue by Antekk27** [Mikadily/poloniexlendingbot#61]
From time to time I have the following effect with the bot:

- I start the bot
- the bot does his job like expect: cancels orders and places new ones.
- the bot goes on running, but is not carrying out any action. No output, no error is showing up.
- Having a look at poloniex lending area, I can see the the BTC Lending Balances is growing up slowly, which means that the bot 
as a matter of fact is not placing any offers.

I tested the bot under windows 7 on a private computer and under debian unix on a virutal server. Both installations shows the same effect from time to time.

Below you can see my config. There are still some DASH and STR loans open. With coinconfig = ["BTC:0.03:1","DASH:0.0:0","STR:0.0:0"] the bot should at least log "maxactive amount for ' + activeCur + ' set to 0, won\'t lend.'" every minute, but it doesn't

Kind Regards from Germany
Andreas

---

```
[BOT]
#sleep between iterations, time in seconds
sleeptime = 60
#minimum daily lend rate in percent
mindailyrate = 0.055
#max rate. 2% is good choice because it's default at margin trader interface. 5% is max to be accepted by the exchange
maxdailyrate = 2
#The number of offers to split the available balance uniformly across the [gaptop, gapbottom] range.
spreadlend = 3
#The depth of lendbook (in percent of lendable balance) to move through before placing the first (gapbottom) and last (gaptop) offer.
#if gapbottom is set to 0, the first offer will be at the lowest possible rate. However some low value is recommended (say 10%) to skip dust offers
gapbottom = 15
gaptop = 100
#Daily lend rate threshold after which we offer lends for 60 days as opposed to 2. If set to 0 all offers will be placed for a 2 day period
sixtydaythreshold = 0.06
#AutoRenew - if set to 1 the bot will toggle the AutoRenew flag for the loans when you stop it (Ctrl+C) and clear the AutoRenew flag when started
autorenew = 0
#custom config per coin, useful when closing positions etc.
#syntax: [COIN:mindailyrate:maxactiveamount, ... COIN:mindailyrate:maxactiveamount]
#if maxactive amount is 0 - stop lending this coin. in the future you'll be able to limit amount to be lent.
coinconfig = ["BTC:0.03:1","DASH:0.0:0","STR:0.0:0"]
#this option creates a json log file instead of console output which includes the most recent status
#uncomment both jsonfile and jsonlogsize to enable
#jsonfile = botlog.json
#limits the amount of log lines to save
#jsonlogsize = 200
#enables a webserver for the www folder, in order to easily use the lendingbot.html with the json log
#startWebServer = true
```

**Comment by Evanito** [comment]
You said this issue happens from time to time, does it fix after a while? And if so, for about how long does the bug last?

**Comment by Antekk27** [comment]
Once the bot stucks, it does not recover. I have to stop and start the bot again. Then it may work for several cycles (doing something and creating console-output according 'sleeptime'). And after a while, the bot stops working again.

I hope this helps.

I thanx to all working on this project for a nice piece of software!!

**Comment by rnevet** [comment]
What are the last log lines you see?

**Comment by Mikadily** [comment]
I think I'm aware of the issue. I think it's due to unreliable network connection - server or client.
I was debugging it, and basically I found out that bot's is stuck on api request. Well I would expect urllib2.Request to time out at some point, but it doesn't. So the only way I could think to fix it, run each cycle in a separate thread, and if the thread is running for to long just terminate it. And I implemented it in separate branch https://github.com/Mikadily/poloniexlendingbot/commits/multithreading

**Comment by Antekk27** [comment]
@rnevet: Sometimes ist shows like:

Sometimes it show like this:

Welcome to Poloniex Lending Bot
2016-05-13 07:50:57 Using custom mindailyrate 0.0300% for BTC

and no further output occurs.

Today it shows like this:
2016-05-13 09:20:43 Using custom mindailyrate 0.0300% for BTC
2016-05-13 09:21:44 Using custom mindailyrate 0.0300% for BTC
2016-05-13 09:21:45 Placing 0.00250573 BTC at 0.03399% for 2 days... Loan order placed.
**2016-05-13 09:21:45 Placing 0.00501146 BTC at 0.03409% for 2 days... Loan order placed.
2016-05-13 09:31:59 Using custom mindailyrate 0.0300% for BTC**
2016-05-13 09:33:01 Using custom mindailyrate 0.0300% for BTC
2016-05-13 09:33:02 Placing 0.02534706 BTC at 0.03649% for 2 days... Loan order placed.
2016-05-13 09:33:02 Placing 0.01267353 BTC at 0.0365% for 2 days... Loan order placed.
2016-05-13 09:34:04 Canceling all BTC orders... Loan offer canceled.
2016-05-13 09:34:04 Canceling all BTC orders... Loan offer canceled.
2016-05-13 09:34:04 Using custom mindailyrate 0.0300% for BTC
2016-05-13 09:34:05 Placing 0.01267353 BTC at 0.0313% for 2 days... Loan order placed.
2016-05-13 09:34:05 Placing 0.01267353 BTC at 0.0337% for 2 days... Loan order placed.
2016-05-13 09:34:05 Placing 0.01267353 BTC at 0.0359% for 2 days... Loan order placed.
Lended: [1.989 BTC @ 0.0499%] [0.184 DASH @ 0.7786%] [146454.948 STR @ 0.6775%]

As you can see in the bold area, today the bot stops working for some minutes an the goes on doing his work.

@Mikadily: Thanx for your hint, I will try out the version from the multithreading branch.

Kind regards
Andreas




To

**Comment by Kurairaito** [comment]
Hi,

Can you try my fix ? I had the same issue several month ago (at first I killed the process automatically if output file wasn't updated for some time and started it again). I now use this for a month without any bot freeze, but don't use my bandwith like when it didn't works.

Regards.

**Comment by rnevet** [comment]
Merged a fix https://github.com/Mikadily/poloniexlendingbot/pull/69, added a 30sec timeout on the socket level.

**Comment by Antekk27** [comment]
Hi there,

thanx for the fix. I am quite busy at the moment, but I will test the fixed version soon.

Regards
Andreas

**Comment by Kurairaito** [comment]
I can confim this fix works fine, I lost connection many hours and many time the past two days, and bot recovered without difficulties.

There are no log in html but python raise an exection as expected : 
```
Traceback (most recent call last):
  File "/volume1/poloBot/lendingbot.py", line 473, in <module>
    refreshTotalLended()
  File "/volume1/poloBot/lendingbot.py", line 223, in refreshTotalLended
    cryptoLended = bot.returnActiveLoans()
  File "/volume1/poloBot/poloniex.py", line 101, in returnActiveLoans
    return self.api_query('returnActiveLoans')
  File "/volume1/poloBot/poloniex.py", line 58, in api_query
    ret = urllib2.urlopen(urllib2.Request('https://poloniex.com/tradingApi', post_data, headers))
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 449, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1240, in https_open
    context=self._context)
  File "/usr/lib/python2.7/urllib2.py", line 1197, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno 101] Network is unreachable>
```

**Comment by denravonska** [comment]
I have seen this as well after a network error (I assume since the result dictionary is empty). It looks like the code handles all exceptions but it gets stuck in limbo anyway. 

`Traceback (most recent call last):
  File "lendingbot.py", line 459, in <module>
    cancelAndLoanAll()
  File "lendingbot.py", line 290, in cancelAndLoanAll
    lendingBalances = bot.returnAvailableAccountBalances("lending")['lending']
  File "/e/coins/poloniexlendingbot-0.2/poloniex.py", line 83, in returnAvailableAccountBalances
    return self.api_query('returnAvailableAccountBalances',{"account":account})
  File "/e/coins/poloniexlendingbot-0.2/poloniex.py", line 58, in api_query
    ret = urllib2.urlopen(urllib2.Request('https://poloniex.com/tradingApi', post_data, headers))
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 449, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1240, in https_open
    context=self._context)
  File "/usr/lib/python2.7/urllib2.py", line 1197, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno 104] Connection reset by peer>
Traceback (most recent call last):
  File "lendingbot.py", line 457, in <module>
    refreshTotalLended()
  File "lendingbot.py", line 230, in refreshTotalLended
    for item in cryptoLended["provided"]:
KeyError: 'provided'
Traceback (most recent call last):
  File "lendingbot.py", line 457, in <module>
    refreshTotalLended()
  File "lendingbot.py", line 230, in refreshTotalLended
    for item in cryptoLended["provided"]:
KeyError: 'provided'
Internal error. Please try again.
You might want to edit config file (default.cfg) and put correct apisecret and key values`

and

> Internal error. Please try again.

I'm not 100% sure that these are what's causing it but I have seen it do nothing after one of these errors.

**Comment by rnevet** [comment]
Well the log suggests that the bot exited on line https://github.com/Mikadily/poloniexlendingbot/blob/master/lendingbot.py#L280 , in cancelAndLoanAll function, worth looking into it.

**Comment by Evanito** [comment]
Following up and closing. Believed to be solved by #100
